### PR TITLE
Create session error code

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

- If the CreateSession failed, the error code we report is the same as a meta request failure, however the CreateSession http response was not reported from the finished callback, as it's not really part of the meta request

*Description of changes:*

- Introduce a new `AWS_ERROR_S3EXPRESS_CREATE_SESSION_FAILED` for CreateSession failed.
- Also, removed `aws_s3_response_status` to use the`AWS_HTTP_STATUS_CODE*`, there is no reason to keep two different set of enums for the status code. It makes me confused.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3_meta_request.c [hunk 1]
```diff
@@ -1088,15 +1088,15 @@ static int s_s3_meta_request_error_code_from_response_status(int response_status
...
+        case AWS_HTTP_STATUS_CODE_206_PARTIAL_CONTENT:
+        case AWS_HTTP_STATUS_CODE_204_NO_CONTENT:
             error_code = AWS_ERROR_SUCCESS;
             break;
-        case AWS_S3_RESPONSE_STATUS_INTERNAL_ERROR:
+        case AWS_HTTP_STATUS_CODE_500_INTERNAL_SERVER_ERROR:
             error_code = AWS_ERROR_S3_INTERNAL_ERROR;
             break;
-        case AWS_S3_RESPONSE_STATUS_SLOW_DOWN:
+        case AWS_HTTP_STATUS_CODE_503_SERVICE_UNAVAILABLE:
```
Comment (waahm7) [inline source/s3_meta_request.c:1099]: Trivial: Maybe add a comment that we are getting throttled, so we should slow down.

File: tests/s3_s3express_client_test.c [hunk 2]
```diff
@@ -590,3 +590,58 @@ TEST_CASE(s3express_client_get_object_multiple) {
     aws_s3_tester_clean_up(&tester);
     return AWS_OP_SUCCESS;
 }
+
+TEST_CASE(s3express_client_get_object_error) {
```
Comment (waahm7) [inline tests/s3_s3express_client_test.c:594]:
```suggestion
TEST_CASE(s3express_client_get_object_create_session_error) {
```

File: source/s3express_credentials_provider.c [hunk 3]
```diff
@@ -314,16 +314,18 @@ static void s_on_request_finished(
...
-
     AWS_LOGF_DEBUG(
         AWS_LS_AUTH_CREDENTIALS_PROVIDER,
         "CreateSession call completed with http status %d and error code %s",
         meta_request_result->response_status,
         aws_error_debug_str(error_code));
 
+    if (error_code) {
+        /* If anything failed from the CreateSession meta request. Provide an CreateSession failure error code to the
+         * user */
```
Comment (waahm7) [inline source/s3express_credentials_provider.c:325]:
```suggestion
        /* If the CreateSession call failed for any reason, return a generic CreateSession failure error. */
```

**Review by waahm7 — APPROVED** [review]

File: source/s3express_credentials_provider.c [hunk 4]
```diff
@@ -314,16 +314,18 @@ static void s_on_request_finished(
     struct aws_credentials *credentials = NULL;
     int error_code = meta_request_result->error_code;
 
-    if (error_code == AWS_ERROR_SUCCESS && meta_request_result->response_status != 200) {
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/s3express_credentials_provider.c:317]: Updated

I am checking the request sent via the error response body to be empty or not.

File: source/s3express_credentials_provider.c [hunk 5]
```diff
@@ -316,14 +316,21 @@ static void s_on_request_finished(
...
 
-    if (error_code) {
-        /* If anything failed from the CreateSession meta request. Provide an CreateSession failure error code to the
-         * user */
+    if (error_code && meta_request_result->error_response_body && meta_request_result->error_response_body->len > 0) {
+        /* The Create Session failed with an error response from S3, provide a specific error code for user. */
         error_code = AWS_ERROR_S3EXPRESS_CREATE_SESSION_FAILED;
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): CreateSession call failed with http status: %d, and error response body is: %.*s",
```
Comment (graebm) [inline source/s3express_credentials_provider.c:329]: Nice, logging the response body 👍

**Review by graebm — APPROVED** [review]

**Pull request merged**